src/crimson/osd/exceptions.h

```diff
@@ -60,3 +60,525 @@ struct input_output_error : public error {
+            maybe_handle_error_t<ErrorVisitorT, Futurator> maybe_handle_error(
+              std::forward<ErrorVisitorT>(errfunc),
+              std::move(future).get_exception()
+            );
+            (maybe_handle_error.template operator()<WrappedAllowedErrorsT>() , ...);
+            return plainify(std::move(maybe_handle_error).get_result());
+          } else {
+            return plainify(Futurator::apply(std::forward<ValueFuncT>(valfunc),
+                                             std::move(future).get()));
+          }
+        })};
+    }
+
+    // taking ErrorFuncOne and ErrorFuncTwo separately from ErrorFuncTail
+    // to avoid SFINAE
+    template <class ValueFunc,
+              class ErrorFuncOne,
+              class ErrorFuncTwo,
+              class... ErrorFuncTail>
+    auto safe_then(ValueFunc&& value_func,
+                   ErrorFuncOne&& error_func_one,
+                   ErrorFuncTwo&& error_func_two,
+                   ErrorFuncTail&&... error_func_tail) {
+      return safe_then(
+        std::forward<ValueFunc>(value_func),
+        composer(std::forward<ErrorFuncOne>(error_func_one),
+                 std::forward<ErrorFuncTwo>(error_func_two),
+                 std::forward<ErrorFuncTail>(error_func_tail)...));
+    }
+
+    template <class Func>
+    void then(Func&&) = delete;
+
+
+  private:
+    // for the sake of `plainify()` let any errorator convert errorated
+    // future into plain one.
+    template <class... AnyWrappedAllowedErrorsT>
+    friend class errorator;
+
+    base_t&& as_plain_future() && {
+      return std::move(*this);
+    }
+  };
+
+  // the visitor that forwards handling of all errors to next continuation
+  struct pass_further {
+    template <class ErrorT>
+    decltype(auto) operator()(ErrorT&&) {
+      using DecayedT = std::decay_t<ErrorT>;
+      static_assert((... || std::is_same_v<WrappedAllowedErrorsT, DecayedT>),
+                    "passing further disallowed ErrorT");
+      return DecayedT::make();
+    }
+  };
+
+  struct discard_all {
+    template <class ErrorT>
+    decltype(auto) operator()(ErrorT&&) {
+      static_assert((... || std::is_same_v<WrappedAllowedErrorsT,
+                                           std::decay_t<ErrorT>>),
+                    "discarding disallowed ErrorT");
+    }
+  };
+
+  template <class ErrorFunc>
+  static decltype(auto) all_same_way(ErrorFunc&& error_func) {
+    return [
+      error_func = std::forward<ErrorFunc>(error_func)
+    ] (auto&& e) mutable -> decltype(auto) {
+      using DecayedT = std::decay_t<decltype(e)>;
+      return DecayedT::handle(std::forward<ErrorFunc>(error_func))(e);
+    };
+  };
+
+  // get a new errorator by extending current one with new error
+  template <class... NewWrappedAllowedErrorsT>
+  using extend = errorator<WrappedAllowedErrorsT...,
+                           NewWrappedAllowedErrorsT...>;
+
+  // comparing errorators
+  template <class ErrorT>
+  struct is_carried {
```

Comment:
Nice snippet. The `is_carried_v` has been reworked a bit since deduplication as it requires counting. Still, the idea with nesting is really cool and will be applied there. Thanks!

Reply:
Switched to `contains_once`.
`::safe_then()` without `ErrorVisitor` specified does `pass_further{}` now.